### PR TITLE
Twister seed

### PR DESCRIPTION
**Pull request by asemjonovs**
Add support to twister to pass seed argument to native posix builds configured with ZTEST_DO_THE_SHUFFLE

### CODE REVIEW

File: scripts/pylib/twister/twisterlib.py
Comment (yperess) [inline scripts/pylib/twister/twisterlib.py:2723]: Why just in native?

File: scripts/twister
Comment (yperess) [inline scripts/twister:912]: I'm by no means an expert in Python, but 0 seems like a valid seed, can we instead (above) set `default=random.randrange(8192)`?

**Review by yperess — CHANGES_REQUESTED** [review]
These commits are all related, can you squash them? I don't believe we want to merge the full history here.

File: scripts/pylib/twister/twisterlib.py
Comment (asemjonovs) [reply scripts/pylib/twister/twisterlib.py:2723]: The seed argument is only valid when configured to use `fake_entropy_native_posix.c`.  Passing seed for any other selected config will cause an error.  The random number generator may use a different entropy module for random numbers.

File: scripts/twister
Comment (asemjonovs) [reply scripts/twister:912]: yeah, 0 can be a valid seed. Let me fix this.

**Comment by nashif** [comment]
Would help if you could pull in the shuffle PR into this until it is merged, this way one would not wonder about the shuffle feature not being present in the tree.

File: scripts/pylib/twister/twisterlib.py
Comment (aescolar) [reply scripts/pylib/twister/twisterlib.py:2724]: native_posix is not the only "native" target type. Upstream nrf52_bsim is also (and others downstream). nrf52_bsim supports --random_seed (or --rs), but not --seed

File: scripts/twister
Comment (aescolar) [inline scripts/twister:690]: By default twister should not switch random seeds for the underlaying tests. It should be a conscious choice by the user to do so. Otherwise a) people will be baffled when tests fail in one run but not the next. b) it adds flakiness to CI where tests fail only some times.

File: scripts/twister
Comment (aescolar) [inline scripts/twister:691]: If this is only used in the native targets, it should say that it is so, and only affecting the native target random generator. Otherwise this would seem to be some twister pseudo-random number generator.

File: scripts/twister
Comment (aescolar) [inline scripts/twister:911]: Silence is golden still. Unless this is going to be used for something, it would be better to not print it. (it won't be used for most targets)

File: scripts/pylib/twister/twisterlib.py
Comment (aescolar) [inline scripts/pylib/twister/twisterlib.py:2726]: Why only propagate the --seed option if the testcase shuffling is selected?

**Review by aescolar — DISMISSED** [review]

File: scripts/twister
Comment (asemjonovs) [reply scripts/twister:691]: Sure, I can update and mention that this is used for native posix only.

File: scripts/pylib/twister/twisterlib.py
Comment (asemjonovs) [reply scripts/pylib/twister/twisterlib.py:2724]: Thanks - will update to reflect `native_posix`

File: scripts/twister
Comment (asemjonovs) [reply scripts/twister:690]: @yperess - any concerns with pushing the seed randomness onto the user?

File: scripts/pylib/twister/twisterlib.py
Comment (aescolar) [reply scripts/pylib/twister/twisterlib.py:2726]: I guess you could check for `CONFIG_FAKE_ENTROPY_NATIVE_POSIX`?
(and it could be up to the user to pass the seed for some other reason(?))

File: scripts/twister
Comment (yperess) [reply scripts/twister:690]: I don't mind at all, Zephyr's CI belongs to them. I personally am in favor of having it done that way this way we can keep finding issues without someone having to manually do it at home, but I 100% understand if Zephyr doesn't want to do it. For what it's worth, Google does this and it's OK, we just say if a failure happens in a PR that doesn't touch it you should just re-run the submit task. Another bot will look at which tests fail and will mark a test as flaky.

File: scripts/twister
Comment (yperess) [reply scripts/twister:911]: The goal is that when a test fails you will have the seed to be able to replay. Though you're right, this can be an error log `if` the test failed.

File: scripts/twister
Comment (hakehuang) [reply scripts/twister:690]: I think we should disable this by default. it may confuse users to reproduce issues.

File: scripts/twister
Comment (aescolar) [inline scripts/twister:689]: If you don't give it a default, it will be `None` when not provided, which would allow you to not need `SEED_INVALID` or `seed_valid`.
```diff
diff --git a/scripts/pylib/twister/twisterlib.py b/scripts/pylib/twister/twisterlib.py
index 8882b91284..54bf59a2e6 100755
--- a/scripts/pylib/twister/twisterlib.py
+++ b/scripts/pylib/twister/twisterlib.py
@@ -486,8 +486,7 @@ class BinaryHandler(Handler):
         self.asan = False
         self.ubsan = False
         self.coverage = False
-        self.seed_valid = False
-        self.seed = 0;
+        self.seed = None;
 
     def try_kill_process_by_pid(self):
         if self.pid_fn:
@@ -565,7 +564,7 @@ class BinaryHandler(Handler):
                        ] + command
             run_valgrind = True
 
-        if self.seed_valid:
+        if self.seed != None:
             command = command + ["--seed="+str(self.seed)];
 
         logger.debug("Spawning process: " +
@@ -2646,8 +2645,8 @@ class ProjectBuilder(FilterBuilder):
                     more_info = "build"
 
                 if ( instance.status in ["error", "failed", "timeout", "flash_error"]
-                     and hasattr(self.instance.handler, 'seed_valid')
-                     and self.instance.handler.seed_valid ):
+                     and hasattr(self.instance.handler, 'seed')
+                     and self.instance.handler.seed != None ):
                     more_info += "/seed: " + str(self.seed)
 
             logger.info("{:>{}}/{} {:<25} {:<50} {} ({})".format(
@@ -2727,11 +2726,10 @@ class ProjectBuilder(FilterBuilder):
             if instance.handler.type_str == "device":
                 instance.handler.suite = self.suite
 
-            if(self.seed != SEED_INVALID and instance.platform.name == "native_posix"):
+            if(self.seed != None and instance.platform.name == "native_posix"):
                 self.parse_generated()
                 if('CONFIG_FAKE_ENTROPY_NATIVE_POSIX' in self.defconfig and
                     self.defconfig['CONFIG_FAKE_ENTROPY_NATIVE_POSIX'] == 'y'):
-                    instance.handler.seed_valid = True
                     instance.handler.seed = self.seed
 
             instance.handler.handle()
diff --git a/scripts/twister b/scripts/twister
index ad81327d0e..8049505574 100755
--- a/scripts/twister
+++ b/scripts/twister
@@ -686,7 +686,7 @@ structure in the main Zephyr tree: boards/<arch>/<board_name>/""")
              "as 'slow' in testcase.yaml. Normally these are only built.")
 
     parser.add_argument(
-        "--seed", type=int, default=SEED_INVALID,
+        "--seed", type=int,
         help="Seed for native posix pseudo-random number generator")
 
     parser.add_argument(
```

File: scripts/pylib/twister/twisterlib.py
Comment (aescolar) [inline scripts/pylib/twister/twisterlib.py:568]: A comment stating this is only used for native_posix would be nice

File: scripts/twister
Comment (asemjonovs) [reply scripts/twister:689]: Thanks @aescolar! This will make it look cleaner.

File: scripts/pylib/twister/twisterlib.py
Comment (yperess) [inline scripts/pylib/twister/twisterlib.py:592]: Can we rename the argument to `--test-seed`?

File: scripts/pylib/twister/twisterlib.py
Comment (yperess) [inline scripts/pylib/twister/twisterlib.py:2763]: I think there's also `native_posix_64` that's valid here. Can we just check that `instance.platform.name` begins with `native_posix`?

File: scripts/pylib/twister/twisterlib.py
Comment (asemjonovs) [reply scripts/pylib/twister/twisterlib.py:592]: This argument is defined in `fake_entropy_native_posix.c`.  If we rename it we may affect others who use this arg as part of their test setup.

**Review by yperess — DISMISSED** [review]

**Comment by yperess** [comment]
@asemjonovs please fix the compliance checks, but LGTM overall

File: subsys/testsuite/ztest/src/ztest_new.c
Comment (yperess) [inline subsys/testsuite/ztest/src/ztest_new.c:766]: Can you move this to another PR. I think we need to be able to re-run multiple tests/suites. Also, keep in mind that the test name doesn't need to be unique, instead it is the combination of suite/test that needs to be unique. I think this extra feature will slow down merging this PR.

**Review by yperess — CHANGES_REQUESTED** [review]

File: subsys/testsuite/ztest/src/ztest_new.c
Comment (asemjonovs) [reply subsys/testsuite/ztest/src/ztest_new.c]: Moved to https://github.com/zephyrproject-rtos/zephyr/pull/44382

**Review by yperess — DISMISSED** [review]

**Comment by yperess** [comment]
Rebased from the "update" UI

File: scripts/pylib/twister/twisterlib.py
Comment (yperess) [inline scripts/pylib/twister/twisterlib.py:91]: nit: just noticed this, I believe that the style guide is to have 2 blank lines before a class

**Review by yperess — DISMISSED** [review]

File: scripts/pylib/twister/twisterlib.py
Comment (asemjonovs) [reply scripts/pylib/twister/twisterlib.py]: Done

**Review by yperess — APPROVED** [review]
@nashif this looks good to me and will help deflake our tests at Google. I just want to make sure you get a chance to review this. I assume that any test that runs with this might cause issues with twister's output parsing? Do we need to update twister to handle seeing the same test running _N_ times? If so, I think that can be done in another PR to keep this one from growing too big.

**Review by yperess — DISMISSED** [review]
Refresh approve after rebase

**Review by yperess — DISMISSED** [review]

**Comment by nashif** [comment]
> @nashif this looks good to me and will help deflake our tests at Google. I just want to make sure you get a chance to review this.

@yperess @asemjonovs what is missing here is documentation and explaining the rationale behind it and how this feature can be used. Please spend some time on documenting all of this, the dependency between ztest and twister and how a test can use this feature.

> Do we need to update twister to handle seeing the same test running _N_ times? If so, I think that can be done in another PR to keep this one from growing too big.

yes, we need to handle the output, it might not be critical in the current version of twister, but will be once we merge the reporting overhaul PR I have been working on.

I will spend some more time on this, but please provide some docs.

**Comment by asemjonovs** [comment]
@nashif, @yperess - I added documentation to ztest.rst.  I didn't feel there was an appropriate place in twister.rst for the seed info.  I feel like the help text in twister script is sufficient, if not please let me know. Please review latest changes.  Thanks!

**Comment by yperess** [comment]
> @nashif, @yperess - I added documentation to ztest.rst. I didn't feel there was an appropriate place in twister.rst for the seed info. I feel like the help text in twister script is sufficient, if not please let me know. Please review latest changes. Thanks!

What about adding a new section next to `Running in Integration Mode`, `Running Tests on Hardware`, `Executing tests on a single device`? You could add something like:

```
Executing tests in random order
========================
Using ztest's `CONFIG_ZTEST_SHUFFLE` it is possible to run your tests our of order.

<explain why it's good>

<explain how to use it + example>
```

**Comment by asemjonovs** [comment]
@yperess Updated documentation in both files.

**Review by yperess — APPROVED** [review]

**Comment by asemjonovs** [comment]
@nashif have you had a chance to review this?

File: scripts/pylib/twister/twisterlib.py
Comment (nashif) [inline scripts/pylib/twister/twisterlib.py:2926]: instead of hardcoding this native_posix, I would add a new attribute (seed: true, or whatever)  to the platform yaml, i.e. `boards/posix/native_posix/native_posix.yaml`  and just check for that here.

File: scripts/pylib/twister/twisterlib.py
Comment (nashif) [inline scripts/pylib/twister/twisterlib.py:2928]: hmm, not sure about that, dont you want to always enable fake entropy when running twister with --seed?

File: scripts/pylib/twister/twisterlib.py
Comment (asemjonovs) [reply scripts/pylib/twister/twisterlib.py:2926]: @nashif - What I'm trying to achieve here is to reduce the parsing of defined configs here to only POSIX to reduce any overhead for other platforms as they will never have `FAKE_ENTROPY_NATIVE_POSIX` defined.  At this point in the script the defined configs are not populated. Adding a new attribute to platform yaml to populate defined configs doesn't feel right to me.  I can use `instance.platform.arch == "posix"` instead which will include the `native_posix_64` platform variant as this is part of the platform yaml.  What are your thoughts?

Comment (nashif) [reply scripts/pylib/twister/twisterlib.py:2926]: actually the data is available at this point,  I am talking about the yaml file for the platform which is one of the first things to be parsed and can be used for this type of checks.

**Review by nashif — APPROVED** [review]

**Pull request merged**
